python/pdo/contract/replication.py

```diff
@@ -0,0 +1,214 @@
+                        use_client = False
+                        break
+            if use_client:
+                self.storage_clients_to_replicate.append(client)
+         
+        # make sure that the number of reliable storage services is sufficient for replication 
+        if len(self.storage_clients_to_replicate) < replication_params['num_provable_replicas']-1: # -1 since we do not explicitly replicate to 
+                    # the sservice associated with the eservice
+            logger.exception('Cannot create Replicator instance: Insufficient number of reliable storage services available for replication')
+            raise ReplicationException('Cannot create Replicator instance: Insufficient number of reliable storage services available for replication')
+    
+    # -----------------------------------------------------------------
+
+    def push_blocks_to_storage_service(self, client, block_data_list, expiration):
+        """Wrapper around client.store_blocks to handle Exceptions during store_blocks. 
+        This is the task function used by a worker-thread that attempts replication to a specific storage service"""
+        
+        try:
+            response = client.store_blocks(block_data_list, expiration)
+            if response is None :
+                Replicator.__queue_of_SSURLS_raising_exceptions__.put(client.ServiceURL) 
+                logger.warn("Adding Storage Service at %s to the set of  bad ones : block_store_list is None", str(client.ServiceURL))
+        except Exception as e:
+            Replicator.__queue_of_SSURLS_raising_exceptions__.put(client.ServiceURL)
+            logger.warn("Adding Storage Service at %s to the set of  bad ones : uknown exception : %s ", str(client.ServiceURL), str(e))
+            response = None
+
+        return response
+    
+    # -----------------------------------------------------------------
+
+    def replicate_change_set(self):
+        """ Replicate change set to the set of reliable storage services."""
+            
+        # identify if replication can be skipped, including the case where there is only one provisioned enclave
+        if len(self.storage_clients_to_replicate) == 0:
+            logger.info('Skipping replication: Only one provisioned enclave, so nothing to replicate')
+            return
+       
+        if len(self.blocks_to_replicate) == 0:
+            logger.info('Skipping replication: No change set, so nothing to replicate')
+            return
+
+        # submit individual replication tasks to various storage services asynchronously. Each replica is handled by a separate thread 
+        executor = concurrent.futures.ThreadPoolExecutor(max_workers = len(self.storage_clients_to_replicate))
+        futures_to_replication_tasks = dict() # to keep track of the tasks
+        expiration = self.availability_duration
+        for index, client in enumerate(self.storage_clients_to_replicate):
```

Comment:
why are you using enumerate here? you aren't using the index anywhere

Reply:
thank you, yes, i have fixed this